src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs

```diff
@@ -0,0 +1,262 @@
+            IReadOnlySet<Symbol> hoistedVariables,
+            IReadOnlyDictionary<Symbol, CapturedSymbolReplacement> nonReusableLocalProxies,
+            SynthesizedLocalOrdinalsDispenser synthesizedLocalOrdinals,
+            VariableSlotAllocator slotAllocatorOpt,
+            int nextFreeHoistedLocalSlot,
+            DiagnosticBag diagnostics)
+            : base(method, methodOrdinal, asyncMethodBuilderMemberCollection, F,
+                  state, builder, hoistedVariables, nonReusableLocalProxies, synthesizedLocalOrdinals,
+                  slotAllocatorOpt, nextFreeHoistedLocalSlot, diagnostics)
+        {
+            Debug.Assert(asyncIteratorInfo != null);
+
+            _asyncIteratorInfo = asyncIteratorInfo;
+            _currentFinallyOrExitLabel = _exprReturnLabel;
+            _exprReturnLabelTrue = F.GenerateLabel("yieldReturn");
+        }
+
+        protected override BoundStatement GenerateSetResultCall()
+        {
+            // ... _exprReturnLabel: ...
+            // ... this.state = FinishedState; ...
+
+            // this.promiseOfValueOrEnd.SetResult(false);
+            // return;
+            // _exprReturnLabelTrue:
+            // this.promiseOfValueOrEnd.SetResult(true);
+
+            // ... _exitLabel: ...
+            // ... return; ...
+
+            return F.Block(
+                // this.promiseOfValueOrEnd.SetResult(false);
+                generateSetResultOnPromise(false),
+                F.Return(),
+                F.Label(_exprReturnLabelTrue),
+                // this.promiseOfValueOrEnd.SetResult(true);
+                generateSetResultOnPromise(true));
+
+            BoundExpressionStatement generateSetResultOnPromise(bool result)
+            {
+                // Produce:
+                // this.promiseOfValueOrEnd.SetResult(result);
+                BoundFieldAccess promiseField = F.Field(F.This(), _asyncIteratorInfo.PromiseOfValueOrEndField);
+                return F.ExpressionStatement(F.Call(promiseField, _asyncIteratorInfo.SetResultMethod, F.Literal(result)));
+            }
+        }
+
+        protected override BoundStatement GenerateSetExceptionCall(LocalSymbol exceptionLocal)
+        {
+            // _promiseOfValueOrEnd.SetException(ex);
+            return F.ExpressionStatement(F.Call(
+                F.Field(F.This(), _asyncIteratorInfo.PromiseOfValueOrEndField),
+                _asyncIteratorInfo.SetExceptionMethod,
+                F.Local(exceptionLocal)));
+        }
+
+        private BoundStatement GenerateJumpToCurrentFinallyOrExit()
+        {
+            Debug.Assert((object)_currentFinallyOrExitLabel != null);
+            return F.If(
+                // if (disposeMode)
+                F.Field(F.This(), _asyncIteratorInfo.DisposeModeField),
+                // goto finallyOrExitLabel;
+                thenClause: F.Goto(_currentFinallyOrExitLabel));
+        }
+
+        BoundStatement AppendJumpToCurrentFinallyOrExit(BoundStatement node)
```

Comment:
>AppendJumpToCurrentFinallyOrExit [](start = 23, length = 32)

Perhaps name this consistently with the method above (`GenerateJumpToCurrentFinallyOrExitBlock`?), or merge the two methods with a `bool asBlock` parameter.